src/cluster_migrateslots.c

```diff
@@ -353,6 +358,151 @@ void fireModuleSlotMigrationEvent(slotMigrationJob *job, int subevent) {
+static slotRange *decodeSlotRange(sds slot_range_str) {
+    int start_slot = atoi(slot_range_str);
+    if (start_slot < 0 || start_slot >= CLUSTER_SLOTS) return NULL;
+
+    while (*slot_range_str && *slot_range_str != '-') slot_range_str++;
+    if (*slot_range_str != '-') return NULL;
+    slot_range_str++; /* Skip '-' */
+
+    int end_slot = atoi(slot_range_str);
+    if (end_slot < 0 || end_slot >= CLUSTER_SLOTS) return NULL;
+    slotRange *slot_range = zmalloc(sizeof(slotRange));
+    slot_range->start_slot = start_slot;
+    slot_range->end_slot = end_slot;
+    return slot_range;
+}
+
+/* Decode a list of slot ranges (separated by " ") from the slot import RDB aux
+ * field.
+ * Returns a list of slotRange* on success, or NULL on failure. */
+list *decodeSlotRangeList(sds slot_range_list_str) {
+    int num_slot_ranges = 0;
+    sds *slot_range_strs = sdssplitlen(slot_range_list_str, sdslen(slot_range_list_str), " ", 1, &num_slot_ranges);
+    if (num_slot_ranges < 1) {
+        sdsfreesplitres(slot_range_strs, num_slot_ranges);
+        return NULL;
+    }
+    list *slot_ranges = createSlotRangeList();
+    for (int j = 0; j < num_slot_ranges; j++) {
+        sds slot_range_str = slot_range_strs[j];
+        slotRange *slot_range = decodeSlotRange(slot_range_str);
+        if (slot_range == NULL) {
+            sdsfreesplitres(slot_range_strs, num_slot_ranges);
+            listRelease(slot_ranges);
+            return NULL;
+        }
+        listAddNodeTail(slot_ranges, slot_range);
+    }
+    sdsfreesplitres(slot_range_strs, num_slot_ranges);
+    return slot_ranges;
+}
+
+/* Decode the slot import RDB aux field.
+ *
+ * See clusterEncodeSlotImportsRDBAuxField for the format. */
+static int clusterDecodeSlotImportsRDBAuxField(int rdbflags, sds s) {
+    if (!server.cluster_enabled || s == NULL) return C_OK;
+
+    if ((rdbflags & RDBFLAGS_AOF_PREAMBLE) != 0) return C_OK;
+
+    list *new_slot_migrations = listCreate();
+    listSetFreeMethod(new_slot_migrations, freeSlotMigrationJob);
+
+    int num_fields = 0;
+    sds *import_fields = NULL;
+    int num_imports = 0;
+    sds *import_strs = sdssplitlen(s, sdslen(s), "/", 1, &num_imports);
+    for (int i = 0; i < num_imports; i++) {
+        import_fields = sdssplitlen(import_strs[i], sdslen(import_strs[i]), ":", 1, &num_fields);
+
+        if (num_fields < 3) goto error;
+        sds job_name = import_fields[0];
+        sds node_name = import_fields[1];
+        sds slot_range_list_str = import_fields[2];
+
+        /* Decode the slot ranges */
+        list *slot_ranges = decodeSlotRangeList(slot_range_list_str);
+        if (!slot_ranges) goto error;
+
+        /* Try to find the corresponding node, or else just insert with no human name. */
+        char *source_node_human_name = NULL;
+        clusterNode *source_node = clusterLookupNode(node_name, CLUSTER_NAMELEN);
+        if (source_node) source_node_human_name = source_node->human_nodename;
```

Comment:
I don't think we should be embedding the source node name here, since that seems somewhat cluster specific and it feels out of place to be embedding it in the RDB. The RDB needs the job ID (Since we couple the slot ranges to the Job ID) and the slots to know which data is being imported.

Reply:
Also - since we are persisting the ESTABLISH command in AOF and in the replication stream - it might lead to slightly odd behavior where:

1. RDB load: No source node
2. AOF load: Yes source node
3. Replication PSYNC: Yes source node
4. Replication full SYNC: No source node

We should align to either "no source node" for all of them or "yes source node" for all of them (the latter is what I have today). To align to "no source node" it means either playing with replication/AOF propagation to not include the source node or ignoring source node during ESTABLISH on replica/AOF load

Reply:
If an old RDB or AOF file is reloaded after restart, do we resume ASM? Is it useful to validate that the source node is the same before resuming in that case?

We don't resume ASM after a failover in the source shard, right?

To me, I keeping the source node seems safer and more explicit. Especially if we require it to be the same after reloading the data from RDB/AOF.

Reply:
Right now - we always give up the ASM after any of the scenarios (RDB load, AOF load, failover). Although in the future, we could attempt to continue it if we include the source node in them